gcsio/src/main/java/com/google/cloud/hadoop/gcsio/InvocationIdGrpcChannelInterceptor.java

```diff
@@ -0,0 +1,57 @@
+package com.google.cloud.hadoop.gcsio;
+
+import io.grpc.CallOptions;
+import io.grpc.Channel;
+import io.grpc.ClientCall;
+import io.grpc.ClientInterceptor;
+import io.grpc.ForwardingClientCall.SimpleForwardingClientCall;
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor;
+import java.util.UUID;
+
+public final class InvocationIdGrpcChannelInterceptor implements ClientInterceptor {
+
+  static final Metadata.Key<String> GOOG_API_CLIENT =
+      Metadata.Key.of("x-goog-api-client", Metadata.ASCII_STRING_MARSHALLER);
+
+  static final String GCCL_INVOCATION_ID = "gccl-invocation-id";
+
+  @Override
+  public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(
+      MethodDescriptor<ReqT, RespT> method, CallOptions callOptions, Channel next) {
+    return new InvocationIdAttachingClientCall<>(next.newCall(method, callOptions));
+  }
+
+  // Non private to avoid synthetic class
+  InvocationIdGrpcChannelInterceptor() {}
+
+  private final class InvocationIdAttachingClientCall<ReqT, RespT>
+      extends SimpleForwardingClientCall<ReqT, RespT> {
+
+    // Non private to avoid synthetic class
+    InvocationIdAttachingClientCall(ClientCall<ReqT, RespT> call) {
+      super(call);
+    }
+
+    @Override
+    public void start(Listener<RespT> responseListener, Metadata headers) {
```

Comment:
why doing this onStart() and not onHeaders()?

Reply:
Using `MetadataUtils.HeaderAttachingClientInterceptor` as a reference which is also used to append headers. Why not using `HeaderAttachingClientInterceptor`? because we want to run our own custom logic about how we want insert invocationIds.

`onHeader` is to intercept the response. Here we are intercepting the request for which `start` is used.